### ISSUE TITLE
How to reverse the array data received from the character.

### ISSUE THREAD

**Issue by SHCPU** [sidorares/node-dbus#85]
I used the fromCharCode () method, but it was clear that the method was not suitable.. Some special data can not resolve.
code:
console.log(String.fromCharCode(110,97,109,101,0,116,101,115,116,0,5,116,121,112,101,0,50,0,5,11,19,0,97,123,115,115,125,27));
result:
nametesttype2
             a{ss}
This previous data does not seem to be a string, and I am very difficult to extract the raw data information.

**Comment by sidorares** [comment]
not quite clear. Where does this data come from? You have zero terminator in the middle, are you sure you need to process all array?

**Comment by SHCPU** [comment]
Yes, I just want to turn to the original data.. The return type of the method is of byte array.
The code of client written by glibc：
_filelist  =  g_variant_new_from_data( G_VARIANT_TYPE( "av" ), client_receive_data,  
            								_size, TRUE, NULL, NULL );  
  
	_iter  =  g_variant_iter_new( _filelist );  

	while( g_variant_iter_next( _iter, "v", &_item ))  
	{  
		gchar  *_name; 
		gchar *_type;
		
		g_variant_lookup( _item, "name", "s", &_name );  
		g_variant_lookup( _item, "type", "s", &_type );  

		g_print( "%s, %s\n", _name, _type );  

The result:
webrt_0.0.2_amd64.deb, 1
webrt_0.0.1_amd64.deb, 1
desktop_components.zip, 1
Downloads, 2
desktop_components, 2

But I don't know how to deal with the client node-dbus's received data.

**Comment by sidorares** [comment]
Answering this properly, ten years on, because the honest answer is two things and only one of them was ever this library's job.

**1. A byte array is a `Buffer`, and that is the easy half.** `String.fromCharCode(...)` is the wrong tool — it builds a JS string from code points, which mangles anything non-ASCII and gives you no way to slice on the NUL boundaries. Today:

```js
const bytes = await proxy.Bytes();
bytes;                      // <Buffer 6e 61 6d 65 00 74 65 73 74 00>
bytes.constructor.name;     // 'Buffer'
bytes.toString('utf8').split('').filter(Boolean);   // [ 'name', 'test' ]
```

Real output. `Buffer` gives you `toString`, `subarray`, `indexOf` and `readUInt32LE` — everything `fromCharCode` was standing in for.

**2. But your data is not a D-Bus value, and that is the actual problem.** Look at what the service is doing:

```c
g_variant_new_from_data(G_VARIANT_TYPE("av"), client_receive_data, _size, TRUE, NULL, NULL);
```

That takes a raw byte buffer and *reinterprets* it as a serialised GVariant. Your bytes confirm it: the strings, then a small offset table, then the type string `a{ss}`, then a framing byte. That is GVariant's own container layout.

**GVariant serialisation and D-Bus wire format are different formats.** They share a type *notation* — `a{ss}` means the same thing in both — but the byte layouts are not interchangeable. So the service is not sending you an `a{ss}`; it is sending an opaque `ay` that happens to contain GLib-serialised bytes. No D-Bus library in any language will decode that for you, because at the D-Bus level it genuinely is just bytes. The glibc client works only because GLib is being handed its own format back.

**The fix is on the service side:** send a real `a{ss}` (or `av`) and let the D-Bus type system carry it. Then it arrives as a plain object, with no decoding at all:

```js
await proxy.Dict();
// { name: 'test', type: '2' }
```

That is real output too, from a service exporting `a{ss}` — and it is exactly what the `g_variant_lookup` loop in your comment is reconstructing by hand.

If you cannot change the service, you would need a GVariant deserialiser, which is out of scope here.

Closing: the library-side question — how to handle the bytes — is answered, and the rest is a format mismatch at the source. Sorry nobody spelled that out in 2015.
